File: av1_poo.py
class Funcionario:
    def __init__(self, nome, matricula, salario_base):
        self.nome = nome
        self.matricula = matricula
        self.__salario_base = salario_base

    def get_salario_base(self):
        return self.__salario_base

    def calcular_salario_final(self):
        return print(f"Salario base é de: {self.get_salario_base()}")

    def set_salario_base(self, novo_salario):

        if novo_salario > 0:
            self.__salario_base = novo_salario
            print(f"Salario de {self.nome} alterado com sucesso para: {self.__salario_base:.2f}")

        else:
            print(f"Salario invalido, Funcionario: {self.nome} | Matricula {self.matricula}, tentativa de alteração para R${novo_salario:.2f}")

class Gerente(Funcionario):
    def __init__(self, nome, matricula, salario_base, bonus_gestao):
        super().__init__(nome, matricula, salario_base)

        self.bonus_gestao = bonus_gestao

    def calcular_salario_final(self):
        salario = super().get_salario_base() + self.bonus_gestao
        print(f"Salario do Gerente {self.nome} é R${salario:.2f}")

class Desenvolvedor(Funcionario):
    def __init__(self, nome, matricula, salario_base, nivel):
        super().__init__(nome, matricula, salario_base)

        self.nivel = nivel

    def calcular_salario_final(self):
        if self.nivel.lower() == 'senior':
            salario = super().get_salario_base() + 1500
            print(f"Salario do Senior {self.nome} é R${salario:.2f}")

        else:
            super().calcular_salario_final()


f1 = Funcionario("Lucas", 123, 4000)
f2 = Gerente("xucas", 456, 5000, 450)
f3 = Desenvolvedor("zucas", 890, 6000, "Senior")

f1.calcular_salario_final()
f1.set_salario_base(7000)
f1.set_salario_base(-7000)

f2.calcular_salario_final()
f3.calcular_salario_final()

        

    
    

    

    
        

File: atividades/Provas/C++/av1_poo.cpp
#include <iostream>
#include <string>
#include <memory>
#include <iomanip>
#include <cmath>

class Funcionario {
private:
    double salario_base;

protected:
    std::string nome;
    std::string matricula;
    
public:
    Funcionario (const std::string& n, const std::string& m, double sb)
        : nome(n), matricula(m), salario_base(sb) {}

    virtual double get_salario_base() const {
        return salario_base;
    }

    virtual void calcular_salario_final() {
        std::cout << "Salario base de " << nome << " e de R$" << std::fixed << std::setprecision(2) << get_salario_base() << "\n";
    }

    void set_salario_base(double novo_salario) {
        if (novo_salario > 0) {
            salario_base = novo_salario;
            std::cout << "O salario de " << nome 
                      << " foi alterado com sucesso para R$" 
                      << std::fixed << std::setprecision(2) << get_salario_base() << "\n";

        } 
        else {
            std::cout << "[ERRO]: Salario invalido, Funcionario: " << nome 
                      << "  | Matricula: " << matricula 
                      << ", tentativa de alteracao para R$" 
                      << std::fixed << std::setprecision(2) << novo_salario << "\n";
        }
    }

    virtual ~Funcionario() = default;
};

class Gerente : public Funcionario {
private:
    double bonus_gestao;

public:
    Gerente (const std::string& n, const std::string& m, double sb, double bg) 
        : Funcionario(n, m, sb), bonus_gestao(bg) {}

    void calcular_salario_final() override {
        double salario = std::fabs(bonus_gestao) + get_salario_base();

        std::cout << "Salario do Gerente " << nome 
                  << " com o bonus de R$" << std::fixed << std::setprecision(2) << bonus_gestao 
                  << " e R$" << std::fixed << std::setprecision(2) << salario << "\n";

    }
};

class Desenvolvedor : public Funcionario {
private:
    std::string nivel;

public:
    Desenvolvedor (const std::string& n, const std::string& m, double sb, const std::string& nv)
        : Funcionario(n, m, sb), nivel(nv) {}

    void calcular_salario_final() override {
        if (nivel == "Senior" || nivel == "senior") {
            double salario = get_salario_base() + 1500;

            std::cout << "Salario do Senior " << nome
                      << " com o bonus de R$1500.00" 
                      << " e R$" << std::fixed << std::setprecision(2) 
                      << salario << "\n";
        }
        else {
            std::cout << "O salario do desenvolvedor " << nome 
                      << " nivel: " << nivel 
                      << " e R$" << std::fixed << std::setprecision(2) << get_salario_base() << "\n";

        }
    }
};

int main() {
    std::unique_ptr<Funcionario> f1 = std::make_unique<Funcionario>("Matheus", "6re54g", 4000.00);
    std::unique_ptr<Funcionario> f2 = std::make_unique<Gerente>("Luiz", "juiujfhbb94", 6000.00, 500.00);
    std::unique_ptr<Funcionario> f3 = std::make_unique<Desenvolvedor>("Pedro", "trhrtyht", 5500.00, "Junior");
    std::unique_ptr<Funcionario> f4 = std::make_unique<Desenvolvedor>("Lucas", "65e4rg2", 50000.00, "Senior");

    std::cout << "\n";

    f1->calcular_salario_final();
    f1->set_salario_base(5500.00);
    f1->calcular_salario_final();
    f1->set_salario_base(-90000.00);
    f1->calcular_salario_final();

    std::cout << "\n";

    f2->calcular_salario_final();
    f2->set_salario_base(8000.00);
    f2->calcular_salario_final();
    f2->set_salario_base(-4600.00);
    f2->calcular_salario_final();

    std::cout << "\n";

    f3->calcular_salario_final();
    f3->set_salario_base(9900.00);
    f3->calcular_salario_final();
    f3->set_salario_base(-80000.00);
    f3->calcular_salario_final();

    std::cout << "\n";

    f4->calcular_salario_final();
    f4->set_salario_base(80000.00);
    f4->calcular_salario_final();
    f4->set_salario_base(-560.00);
    f4->calcular_salario_final();

    std::cout << "\n";

    return 0;
}
